### ISSUE TITLE
Model Loader Class

### ISSUE THREAD

**Issue by lch32111** [lch32111/ChanGameEngine#5]
Now I'm gonna develop my model structure.
So This issue will include all of my thoughts about it.

As a game engine, I'd like to provide a feature to load various 3D object files, such as obj, stl, fbx, and so on.
So, I'm gonna use Assimp library for supporting lots of 3D file types.

In my codes, What I conceived for model structure is like

Model Type
1) Color Material(No Texture Used) : Render GLPrimitive(Triangle, Tetrahedron, Box, Sphere, and so on).
       - bool IsUseTangent; This will be used for telling whether this model uses normal/parallax 
                                         mapping or not
       - There is GLPrimitiveUtil header file to render primitive. I'm gonna convert this file into the one, 
         which feeds the vertex, normal, texCoords, tangents information to the Model Data.
         So I have to add the feature of calculating vertex, normal, texCoords, tangents for each primitive.
        And Before doing this, I think I need to set the types of primitives ahead.

So Maybe It will be good like this

```
Class TriangleVertexGenerator;
Class NormalGenerator;
Vertices* vLocalPointer = TriangleVertexGenerator.getLocalVertexPointer();
Normal* nLocalPointer = TriangleVertexGenerator.getLocalNormalPointer();
TexCoords* nTexPointer = TriangleVertexGenerator.getLocalTexPointer(); 
Tangent* nTanPointer = TriangleVertexGenerator.getLocalTanPointer();
```

Because, the data of triangle will not be changed normally after one instance.
So, I can reuse that kinds of model data from one class.
And I think After implementing more complex object like cylinder, I need to make methods for adjusting the parameters like:

```
Class CylinderVertexGenerator;
CylinderVertexGenerator.setTopRadius(~);
CylinderVertexGenerator.setBottomRadius(~);
CylinderVertexGenerator.generateData();
```

2) Light Map Material
       - bool IsUseTangent; This will be used for telling whether this model uses normal/parallax 
                                         mapping or not
       - I will also use GLPrimitiveUtil like the explanation above.
       - The most important feature of this type is using Assimp library to load lots of models on the 
          internet, because I can't design my own models now. So, I will totally make use of Assimp library.
         But When I learned it from learnopengl.com, the structure of model using assimp library is not that 
         good for me. So, I'm gonna convert it, fitted in my taste and engine. 
         Before going further, I'd like to put a kind of notification of future feature which I want to have : 
         First. Instanced Rendering with model
         Second. converting texture of model with mouse picking from parts to the whole body.
        The two of features which i will implement later should be considered when I start to construct 
        model class.  So, I'm gonna post on the below comment what I thought when I learned from 
        instancing rendering using assimp model from learnopengl.com.

3) Common
       - I have to add the feature of accessing all of the information, vertices, indices, texCoords, tangents 
         for future features.

**Comment by lch32111** [comment]
learnopengl.com에 있는 assimp를 이용한 model class를 이용하면서 알아내는 것.
이걸 보고 Instanced Rendering할 때 고려하면서 클래스를 짜야한다.


이 코드 구조에서 Instancing을 더 개선할 수 있는 방법을 알아냈다.

코드를 읽다보니 발견한 건데

현재 

Model Class Instance Draw
{
	for()
	{
		mesh.instancedraw
	}
}

요런식으로 하고있다. 그렇기 때문에 mesh에 포함된 것만 각각, instance draw를 하고 있는데, 
Model class에는 std::vector<mesh> 이게 있다. 그리고
Mesh class에는 std::vector<Vertex> std::vector<unsigned>(index), std::vector<Texture>로 구성되어있다.

따라서, Model Class를 구성할 때, std::vector<mesh> 모든 데이터를 한 번에 접근할 수 있는 배열로 만들자는 것이다.
즉, 메모리 구조가

```
int verticeNumb = 0;
int indicesNumb = 0;
int textureNumb = 0;
for(int i = 0 ; i < m_meshes.size(); ++i)
{
	verticeNumb += m_meshes[i].m_vertices.size();
	indicesNumb +=  m_meshes[i].m_indices.size();
	textureNumb += m_textures[i].m_textures.size();
}

float* ModelAllVertex = new float[verticesNumb];
float* ModelAllIndices = new float[indicesNumb];
float* ModelAllTexture = new float[textureNumb];

int vAddress = 0;
int iAddress = 0;
int tAddress = 0;
for(int i = 0; i < m_meshes.size(); ++i)
{
	// 내 생각에 이 코드가 제대로 작동하는지 모르겠다. 내 아이디어는
	// ModelAllVertex에 이렇게 순차적으로 각 mesh의 m_Vertices 주소를 넣으면 자동으로 연결되는거.
	// 근데 내 생각에는 이게 잘 될거라고 생각이 들지 않는다.
	// 그래서, 일일이 복사하거나, Assimp 라이브러리를 참고하자.
	// 내 생각엔 이미 Assimp 라이브러리에 모든 Vertex, Indices, Texture가 정리되어 있을 것이다.
	// 그래서 그것에 접근할 수 있게 해서 한 번에 그리도록 하자.
	ModelAllVertex[vAddress] = &(m_meshes[i].m_vertices);
	vAddress += m_meshes[i].m_vertices.size();
	ModelAllIndices[iAddress] = &(m_meshes[i].m_indices);
	iAddress += m_meshes[i].m_indices.size();
	ModelAllTexture[tAddress] = &(m_meshes[i].m_textures);
	tAddress += m_meshes[i].m_textures.size();
}
```

어쩃든, 
float* ModelAllVertex
float* ModelAllIndices
float* ModelAllTexture

요놈들을 통해서, 모델 전체에 대해서 Instance Draw를 하게 되면 엄청 빠를 것이다.
Mesh각각에 대해서 Instance Draw를 하는 것보다

**Comment by lch32111** [comment]
Plus!

Before Going further, I forgot about Parallax Mapping. So, After studying it, I have to set the structure of my Deferred Rendering Shaders for Normal/Parallax Mapping.

In case of Normal Mapping, It is just need to convert G-Buffer shader for normal mapping. And that's it.
However, I think I may add something for Parallax Mapping. So, After learning it, I will write something about structure of shaders where I can use Parallax Mapping.

**Comment by lch32111** [comment]
Tangent Calculation Reference

[http://www.terathon.com/code/tangent.html](url)
[http://www.sunandblackcat.com/tipFullView.php?l=eng&topicid=8&topic=Tangent-Bitangent-Vectors](url)
[https://gamedev.stackexchange.com/questions/68612/how-to-compute-tangent-and-bitangent-vectors](url)
[https://github.com/assimp/assimp/blob/master/code/CalcTangentsProcess.cpp](url)

**Comment by lch32111** [comment]
Shader...

```
// shadertype=glsl
#version 330 core

layout (location = 0) in vec3 aPos;
layout (location = 1) in vec3 aNormal;
layout (location = 2) in vec2 aTexCoords;
layout (location = 3) in vec3 aTangent;

uniform mat4 projection;
uniform mat4 view;
uniform mat4 model;
uniform mat3 ModelNormalMatrix;
uniform bool IsUseTangentSpace;

out vec3 FragPos;
out vec3 Normal;
out vec2 TexCoords;

void main()
{
	gl_Position = projection * view * model * vec4(aPos, 1.0);

	// Lighting Calculation in World Space
	FragPos = vec3(model * vec4(aPos, 1.0));
	TexCoords = aTexCoords;

	if(IsUseTangentSpace)
	{
		// Gram-Schmidt Process
		vec3 T = normalize(vec3(model * vec4(aTangent, 0.0)));
		vec3 N = normalize(vec3(model * vec4(aNormal, 0.0)));
		T = normalize(T - dot(T, N) * N);
		vec3 B = cross(N, T);
		Normal = mat3(T,B,N) * aNormal;
	}
	else
	{
		Normal = ModelNormalMatrix * aNormal;
	}
}
```

대충 이런식으로 해서, 하면 Normal Mapping은 된다.
근데 Parallax Mapping을 다시 공부해서, 이 구조로 가도 되는지 확신을 해보자.

**Comment by lch32111** [comment]
Parallax Mapping을 공부했다. Parallax Mapping은 Tangent Space에서 알고리즘을 실행한다.
Parallax Mapping의 결과물은 새로운 TexCoords이다.

그리고 그 새로운 TexCoords에 따라, NormalMap을 샘플링하여 Normal Mapping까지 한다.

그 때 필요한 Tagent Space 벡터들은 
- viewDir(Fragment -> CameraPos) for Parallax Occlusion Mapping
- lightDir(Fragment -> LightPos) for Self-Shadowing with Parallax Occlusion Mapping

우리는 Deferred Shader의 첫 번째 Pass인 G-Buffer Pass에서 이 계산을 하는게 좋을 것 같다.
근데 Self-Shadowing을 구현 하려면 light info가 필요하니, ParallaxHeightMap 을 Deferred Shade의 두 번째 pass에 올려놓아야 할 것이다.

따라서, 구현은 다음과 같아진다.

```
// shadertype=glsl
#version 330 core

layout (location = 0) in vec3 aPos;
layout (location = 1) in vec3 aNormal;
layout (location = 2) in vec2 aTexCoords;
layout (location = 3) in vec3 aTangent;

uniform mat4 projection;
uniform mat4 view;
uniform mat4 model;
uniform mat3 ModelNormalMatrix;
uniform bool IsUseTangentSpace;

out vec3 FragPos;
out vec3 Normal;
out vec2 TexCoords;
out mat3 TBNmat;

void main()
{
	gl_Position = projection * view * model * vec4(aPos, 1.0);

	// Lighting Calculation in World Space
	FragPos = vec3(model * vec4(aPos, 1.0));
	TexCoords = aTexCoords;

	if(IsUseTangentSpace)
	{
		// Gram-Schmidt Process
		vec3 T = normalize(vec3(model * vec4(aTangent, 0.0)));
		vec3 N = normalize(vec3(model * vec4(aNormal, 0.0)));
		T = normalize(T - dot(T, N) * N);
		vec3 B = cross(N, T);
		TBNmat = mat3(T, B, N);
	}
	else
	{
		Normal = ModelNormalMatrix * aNormal;
	}
}
```

```
// shadertype=glsl

// Deferred First Pass Fragment Shader
// The lighting will be calculated in World Space.

#version 330 core

layout (location = 0) out vec3 gPosition;
layout (location = 1) out vec3 gNormal;
layout (location = 2) out vec4 gAlbedoSpec;
layout (location = 3) out vec3 gEmissive;
layout (location = 4) out vec4 gBool;

in vec3 FragPos;
in vec3 Normal;
in vec2 TexCoords;
in mat3 TBNmat;
in bool IsUseTangent;

struct Material
{
    bool CMorLM; // CM == false, LM == true
    bool isLMdiffuse;
    bool isLMspecular;
    bool isLMemissive;

    // Color Material Component
    vec3 CMambient;
    vec3 CMdiffuse;
    vec3 CMspecular;
    float CMshininess;

    // Light Map Material Component
    sampler2D LMdiffuse;
    sampler2D LMspecular;
    sampler2D LMemissive;

   // Normal/Parallax Mapping
    bool isNormalMap;
    bool isDepthMap;
    sampler2D normalMap;
    sampler2D depthMap;

};

uniform Material material;
uniform vec3 cameraPos;

void main()
{
    // store the fragment position vector in the first gbuffer texture
    gPosition = FragPos;

    if(isDepthMap)
    {
        // Parallax Mapping -> get new TexCoords;
        TexCoords = ParallaxOcclusionMapping(TexCoords, transpose(TBNmat) * (cameraPos - FragPos));
    }

    if(isNormalMap)
    {
        // Normal Mapping -> get new normal, and then transform it into world space.
        Normal = texture(normalMap, TexCoords).xyz;
        Normal = normalize(Normal * 2.0 - 1.0);
        Normal = TBNmat * Normal;
    }

	
    // also store the per-fragment normals into the gbuffer
    gNormal = normalize(Normal);
    
    gBool.a = material.CMorLM ? 0.9 : 0.1; 
    if(material.CMorLM)
    {
        gBool.rgb = vec3(1.0);
        // Light Map
        if(material.isLMdiffuse)	gAlbedoSpec.rgb = texture(material.LMdiffuse, TexCoords).rgb;
        else    gBool.r = 0;   
        if(material.isLMspecular)    gAlbedoSpec.a = texture(material.LMspecular, TexCoords).r;
        else    gBool.g = 0;
        if(material.isLMemissive)    gEmissive.rgb = texture(material.LMemissive, TexCoords).rgb;
        else    gBool.b = 0;
    }
    else
    {
        // Color Material
        gBool.rgb = material.CMambient;
        gAlbedoSpec.rgb = material.CMdiffuse;
        gAlbedoSpec.a = material.CMshininess;
        gEmissive.rgb = material.CMspecular;
    }
}
```

그런데 만약 Self-Shadowing까지 하려면, TBNmatrix가 또 필요하게 된다.
왜냐하면 lightDir를 할 때, 그것을 tangent space로 바꿔야 하기 때문이다.

일단은 이 경우에는 어떻게 해야할지 아직 생각이 나지 않는다.
그래서, 일단은 self shadowing은 보류하고, 일단 이렇게 개발하도록 한다.

self-shadowing을 하려면 Second pass에서 depthmap texture도 보내야 된다.
그래서 이것도 어떻게 해결할지 생각해볼 필요가 있다.

**Comment by lch32111** [comment]
이제 Model Class를 만들기 위해, 여러 자료들을 읽어 보았다.
그것들을 읽은 후에 내가 내린 결론은,

한 모델 데이터를 불러올 때, 여러 Mesh 데이터로 분절이 되는데,
나는 그 Mesh 데이터들을 다시 하나의 버퍼로 통합할려고 했다.

이것은 물론 나는 아주 괜찮은 아이디어이고, DrawCall을 줄일 수 있는 좋은 방법이라고 생각했다.
하지만, Texture를 써야하는 상황이라면 문제가 달라진다.

예를들어 큰 오브젝트가 있다고 해보자. 그리고 그 각 큰 오브젝트는 여러 메쉬로 분리되어있고
여러 텍스쳐가 사용된다고 해보자. 그렇다면, 만약 그 텍스쳐가 OpenGL의 Texxture Samping 허용갯수를 넘어버린다면?

그러면 적절히 렌더링을 할 수 없게 된다. 

만약 Texture를 쓸 필요가 없다고 한다면, 내가 말한 통합 버퍼로 만들어서 한 번에 draw call을 날리는 것이 가능하다. 하지만, 그 통합 버퍼가 차지하는 메모리가 엄청 커서, GPU에 올리지 못한다면? 물론 그러한 모델을 렌더링하려고 하는 것은 좋지 않은 선택이다.

그래서 결과적으로 이전에 LearnOpenGL.com 에서 했던 대로 Mesh를 분절하고 이전에 썼던 InstancedRendering도 그대로 적용하도록 해야겠다.

이렇게 한 Model data를 여러 개의 Mesh로 나누는 것의 장단점을 이제 이해할 수 있다.

장점 : 각 Mesh별로 접근할 수 있어서, 그 각 메쉬에 대해 통제성을 가질 수 있다. 예를들어 대표적으로 Animation이 Bone Mesh별로 이렇게 할 것이다.

단점 : 한 번에 통합되어있지 않아 여러 번의 Draw Call을 날려야 한다.

하지만 컴퓨터 성능으로 인해, 이러한 단점은 엄청나게 많이 되지 않는 이상 괜찮을 거라 본다.
나중에 모델 데이터에 LOD를 적용할 수 있게 한다면, 삼각형 데이터를 더 적게 그려서 그 단점을 조금이나마 낮출 수 있을 것이다.

이제 CGModel 클래스를 작성해보자.

**Comment by lch32111** [comment]
나는 현재 나의 growableArray를 만들지 않아서, stl vector를 쓰고 있지만, 나중에 이것을 위해서 growableArray를 만들어서 메모리 관리를 더 잘할 수 있도록 하자.

**Comment by lch32111** [comment]
I did it now. The model is not using normal/depth map now.
So, I need to test the model to use normal/depth map.
Anyway, This is just the first try. I want to record it.

After testing the normal/depth map, I will upload other records.

![4](https://user-images.githubusercontent.com/19434311/51239468-d46d7500-19bc-11e9-885b-62f8d09a4240.PNG)

**Comment by lch32111** [comment]
I found one task to do :

modify the structure to draw shadow map.
I tried to shadow cast using directional light.
But My structure is using primitive box now.
So I need to change the structure of rendering shadow map

**Comment by lch32111** [comment]
debugged for two days. the bug was related to shader code of parallax mapping. I didn't normalize view direction vector! Because of this I didn't get any crash point in my cpu code. That's the reason why I took two days to debug it. If you don't get any crash point, see shader!..

For me, this kind of bug was the first experience, except for that I went through the same symptom in shadertoy. Maybe I guess it's shader stack overflow?? Anyway, My computer was freezed for almost one minute, and then the shadertoy told me the program exit.

In the case of visual studio, My computer was also freezed for almost one minute, and then I got the message : nvoglv.dll unhandled exception. Fatal Eror program exit ~~ from Visual Studio. The next time I have this kind of symptom, I will look at shader.
